File: .skills/openclaw-skills/skills/globalcaos/teams-hack/scripts/teams.mjs
#!/usr/bin/env node
/**
 * teams.mjs — Microsoft Teams via Graph API
 * 
 * Shares the same MSAL refresh token as the Outlook skill.
 * Token extracted once from Teams localStorage, auto-refreshes for 90+ days.
 * 
 * Subcommands:
 *   token extract-browser    — Print JS to run in Teams tab (extracts refresh token)
 *   token store --refresh-token <token> [--tenant-id <id>]
 *   token test
 *   chats [--top N]                         — List recent chats
 *   chat <chatId> [--top N]                 — Read messages from a chat
 *   chat-send <chatId> --message <text>     — Send a message to a chat
 *   channels <teamId>                       — List channels in a team
 *   channel <teamId> <channelId> [--top N]  — Read channel messages
 *   channel-send <teamId> <channelId> --message <text>
 *   teams                                   — List joined teams
 *   presence [--user-id <id>]               — Get presence status
 *   users [--search X] [--top N]            — Search users in org
 *   me                                      — Profile info
 *   activity [--top N]                      — Activity feed
 *   calendar [--days N]                     — Calendar (same as outlook)
 *   search <query>                          — Search messages across Teams
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync, chmodSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';

const CREDS_DIR = join(homedir(), '.openclaw/credentials');
const TOKEN_FILE = join(CREDS_DIR, 'outlook-msal.json'); // Shared with Outlook!
const GRAPH = 'https://graph.microsoft.com/v1.0';
const GRAPH_BETA = 'https://graph.microsoft.com/beta';

mkdirSync(CREDS_DIR, { recursive: true });

// ─── Token Management (shared with Outlook) ───
function loadCreds() {
  if (!existsSync(TOKEN_FILE)) return {};
  return JSON.parse(readFileSync(TOKEN_FILE, 'utf8'));
}

function saveCreds(creds) {
  writeFileSync(TOKEN_FILE, JSON.stringify(creds, null, 2), { mode: 0o600 });
  try { chmodSync(TOKEN_FILE, 0o600); } catch {}
}

let _accessToken = null;

async function refreshAccessToken() {
  const creds = loadCreds();
  const clientId = creds.client_id || '5e3ce6c0-2b1f-4285-8d4b-75ee78787346';
  const tenantId = creds.tenant_id || 'common';
  const origin = creds.origin || 'https://teams.cloud.microsoft';
  const scope = creds.scope || 'https://graph.microsoft.com/.default offline_access';

  if (creds.access_token && creds.access_token !== 'pending' && creds.expires_at) {
    const expiresAt = new Date(creds.expires_at);
    if (expiresAt > new Date(Date.now() + 5 * 60 * 1000)) {
      _accessToken = creds.access_token;
      return _accessToken;
    }
  }

  const rt = creds.refresh_token;
  if (!rt) throw new Error('No refresh token. Run: teams token extract-browser');

  const tokenUrl = `https://login.microsoftonline.com/${tenantId}/oauth2/v2.0/token`;
  const body = new URLSearchParams({
    client_id: clientId,
    grant_type: 'refresh_token',
    refresh_token: rt,
    scope,
  });

  const resp = await fetch(tokenUrl, {
    method: 'POST', body,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded', 'Origin': origin },
  });
  const data = await resp.json();
  if (data.error) throw new Error(`Token refresh failed: ${data.error_description || data.error}`);

  _accessToken = data.access_token;
  const updated = { ...creds };
  updated.access_token = data.access_token;
  if (data.refresh_token) updated.refresh_token = data.refresh_token;
  updated.expires_at = new Date(Date.now() + (data.expires_in || 3600) * 1000).toISOString();
  updated.updated_at = new Date().toISOString();
  saveCreds(updated);
  return _accessToken;
}

async function getToken() {
  if (_accessToken) return _accessToken;
  return refreshAccessToken();
}

// ─── Graph API ───
async function graphGet(path, beta = false, retried = false) {
  const token = await getToken();
  const base = beta ? GRAPH_BETA : GRAPH;
  const url = path.startsWith('http') ? path : `${base}${path}`;
  const resp = await fetch(url, {
    headers: { 'Authorization': `Bearer ${token}`, 'Content-Type': 'application/json' },
  });
  if (resp.status === 401 && !retried) {
    _accessToken = null;
    await refreshAccessToken();
    return graphGet(path, beta, true);
  }
  if (!resp.ok) throw new Error(`Graph ${resp.status}: ${await resp.text()}`);
  return resp.json();
}

async function graphPost(path, body, beta = false, retried = false) {
  const token = await getToken();
  const base = beta ? GRAPH_BETA : GRAPH;
  const url = path.startsWith('http') ? path : `${base}${path}`;
  const resp = await fetch(url, {
    method: 'POST',
    headers: { 'Authorization': `Bearer ${token}`, 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (resp.status === 401 && !retried) {
    _accessToken = null;
    await refreshAccessToken();
    return graphPost(path, body, beta, true);
  }
  if (resp.status === 204 || resp.status === 201 || resp.status === 202) {
    const text = await resp.text();
    return text ? JSON.parse(text) : { status: 'ok' };
  }
  if (!resp.ok) throw new Error(`Graph ${resp.status}: ${await resp.text()}`);
  return resp.json();
}

// ─── Helpers ───
function parseArgs(argv) {
  const result = { _: [], flags: {} };
  let i = 0;
  while (i < argv.length) {
    if (argv[i].startsWith('--')) {
      const key = argv[i].slice(2);
      if (i + 1 < argv.length && !argv[i + 1].startsWith('--')) {
        result.flags[key] = argv[++i];
      } else {
        result.flags[key] = true;
      }
    } else {
      result._.push(argv[i]);
    }
    i++;
  }
  return result;
}

function out(data) {
  console.log(typeof data === 'string' ? data : JSON.stringify(data, null, 2));
}

function formatMsg(m) {
  return {
    id: m.id,
    from: m.from?.user?.displayName || m.from?.application?.displayName || '?',
    date: m.createdDateTime?.slice(0, 19).replace('T', ' '),
    body: m.body?.content?.replace(/<[^>]*>/g, '').replace(/\s+/g, ' ').trim().slice(0, 500),
    type: m.messageType,
    importance: m.importance,
    hasAttachments: (m.attachments?.length || 0) > 0,
  };
}

// ─── Commands ───

async function cmdTokenExtractBrowser() {
  console.log(`// Run this in the browser console on teams.cloud.microsoft:
// Or let the agent run it via browser(action=act, evaluate)

(() => {
  const keys = Object.keys(localStorage).filter(k => 
    k.includes('refreshtoken') || k.includes('RefreshToken')
  );
  if (keys.length === 0) return { error: 'No refresh tokens found in localStorage' };
  
  const results = keys.map(k => {
    try {
      const parsed = JSON.parse(localStorage.getItem(k));
      return {
        key: k,
        secret: parsed.secret,
        home_account_id: parsed.home_account_id,
        environment: parsed.environment,
        credential_type: parsed.credential_type,
        client_id: parsed.client_id,
      };
    } catch(e) {
      return { key: k, raw: localStorage.getItem(k)?.slice(0, 100) };
    }
  });
  
  // Also get tenant info
  const accountKeys = Object.keys(localStorage).filter(k => {
    try { const v = JSON.parse(localStorage.getItem(k)); return v?.tenantId; } catch { return false; }
  });
  let tenantId = null;
  for (const k of accountKeys) {
    try { tenantId = JSON.parse(localStorage.getItem(k)).tenantId; break; } catch {}
  }
  
  return { tokens: results, tenantId };
})();`);
}

async function cmdTokenStore(args) {
  const rt = args.flags['refresh-token'];
  if (!rt) { console.error('Usage: teams token store --refresh-token <token> [--tenant-id <id>]'); process.exit(1); }
  
  const tenantId = args.flags['tenant-id'] || loadCreds().tenant_id || 'common';
  const creds = {
    client_id: '5e3ce6c0-2b1f-4285-8d4b-75ee78787346',
    tenant_id: tenantId,
    refresh_token: rt,
    origin: 'https://teams.cloud.microsoft',
    scope: 'https://graph.microsoft.com/.default offline_access',
    api: 'graph',
    updated_at: new Date().toISOString(),
  };
  saveCreds(creds);
  console.error('✅ Token saved. Testing...');
  try {
    await refreshAccessToken();
    const me = await graphGet('/me?$select=displayName,mail');
    out({ status: 'ok', user: me.displayName, email: me.mail });
  } catch (e) {
    console.error(`❌ ${e.message}`);
    process.exit(1);
  }
}

async function cmdTokenTest() {
  const me = await graphGet('/me?$select=displayName,mail,userPrincipalName');
  out({ status: 'ok', user: me.displayName, email: me.mail || me.userPrincipalName });
}

async function cmdMe() {
  const me = await graphGet('/me?$select=displayName,mail,userPrincipalName,jobTitle,department,officeLocation,mobilePhone,businessPhones');
  out(me);
}

async function cmdTeams() {
  const data = await graphGet('/me/joinedTeams?$select=id,displayName,description');
  out((data.value || []).map(t => ({
    id: t.id,
    name: t.displayName,
    description: t.description,
  })));
}

async function cmdChats(args) {
  const top = parseInt(args.flags.top) || 20;
  const data = await graphGet(`/me/chats?$top=${top}&$expand=members($select=displayName)&$orderby=lastMessagePreview/createdDateTime desc&$select=id,topic,chatType,lastMessagePreview,createdDateTime`);
  out((data.value || []).map(c => ({
    id: c.id,
    type: c.chatType,
    topic: c.topic || (c.members?.map(m => m.displayName).join(', ')),
    lastMessage: c.lastMessagePreview?.body?.content?.replace(/<[^>]*>/g, '').slice(0, 100),
    lastDate: c.lastMessagePreview?.createdDateTime?.slice(0, 19).replace('T', ' '),
    members: c.members?.map(m => m.displayName),
  })));
}

async function cmdChatMessages(args) {
  const chatId = args._[0];
  if (!chatId) { console.error('Usage: teams chat <chatId> [--top N]'); process.exit(1); }
  const top = parseInt(args.flags.top) || 30;
  
  const data = await graphGet(`/me/chats/${chatId}/messages?$top=${top}&$orderby=createdDateTime desc`);
  out((data.value || []).map(formatMsg));
}

async function cmdChatSend(args) {
  const chatId = args._[0];
  const message = args.flags.message;
  if (!chatId || !message) { console.error('Usage: teams chat-send <chatId> --message <text>'); process.exit(1); }
  
  const result = await graphPost(`/me/chats/${chatId}/messages`, {
    body: { content: message, contentType: 'text' },
  });
  out({ status: 'sent', id: result.id, date: result.createdDateTime });
}

async function cmdChannels(args) {
  const teamId = args._[0];
  if (!teamId) { console.error('Usage: teams channels <teamId>'); process.exit(1); }
  
  const data = await graphGet(`/teams/${teamId}/channels?$select=id,displayName,description,membershipType`);
  out((data.value || []).map(c => ({
    id: c.id,
    name: c.displayName,
    description: c.description,
    type: c.membershipType,
  })));
}

async function cmdChannelMessages(args) {
  const teamId = args._[0];
  const channelId = args._[1];
  if (!teamId || !channelId) { console.error('Usage: teams channel <teamId> <channelId> [--top N]'); process.exit(1); }
  const top = parseInt(args.flags.top) || 20;
  
  const data = await graphGet(`/teams/${teamId}/channels/${channelId}/messages?$top=${top}`);
  out((data.value || []).map(formatMsg));
}

async function cmdChannelSend(args) {
  const teamId = args._[0];
  const channelId = args._[1];
  const message = args.flags.message;
  if (!teamId || !channelId || !message) {
    console.error('Usage: teams channel-send <teamId> <channelId> --message <text>');
    process.exit(1);
  }
  
  const result = await graphPost(`/teams/${teamId}/channels/${channelId}/messages`, {
    body: { content: message, contentType: 'text' },
  });
  out({ status: 'sent', id: result.id, date: result.createdDateTime });
}

async function cmdPresence(args) {
  const userId = args.flags['user-id'];
  const path = userId ? `/users/${userId}/presence` : '/me/presence';
  const data = await graphGet(path);
  out({
    availability: data.availability,
    activity: data.activity,
    statusMessage: data.statusMessage?.message?.content,
  });
}

async function cmdUsers(args) {
  const top = parseInt(args.flags.top) || 25;
  const search = args.flags.search;
  
  let url = `/users?$top=${top}&$select=id,displayName,mail,jobTitle,department,officeLocation`;
  if (search) url += `&$search="displayName:${search}"&$header=ConsistencyLevel:eventual`;
  
  // Search requires ConsistencyLevel header
  const token = await getToken();
  const headers = {
    'Authorization': `Bearer ${token}`,
    'Content-Type': 'application/json',
  };
  if (search) headers['ConsistencyLevel'] = 'eventual';
  
  const resp = await fetch(`${GRAPH}${url}`, { headers });
  if (!resp.ok) throw new Error(`Graph ${resp.status}: ${await resp.text()}`);
  const data = await resp.json();
  
  out((data.value || []).map(u => ({
    id: u.id,
    name: u.displayName,
    email: u.mail,
    title: u.jobTitle,
    department: u.department,
    office: u.officeLocation,
  })));
}

async function cmdActivity(args) {
  const top = parseInt(args.flags.top) || 20;
  // Activity feed requires beta API
  try {
    const data = await graphGet(`/me/teamwork/installedApps?$expand=teamsApp`, true);
    out({ note: 'Activity feed listed installed apps', count: data.value?.length });
  } catch {
    // Fallback: list recent notifications
    console.error('Activity feed requires specific permissions. Showing recent chats instead.');
    await cmdChats({ _: [], flags: { top: String(top) } });
  }
}

async function cmdCalendar(args) {
  const days = parseInt(args.flags.days) || 7;
  const start = new Date().toISOString();
  const end = new Date(Date.now() + days * 86400000).toISOString();
  
  const data = await graphGet(`/me/calendarView?startDateTime=${start}&endDateTime=${end}&$top=50&$select=subject,start,end,location,organizer,isAllDay,isCancelled,showAs,isOnlineMeeting,onlineMeeting&$orderby=start/dateTime`);
  out((data.value || []).map(e => ({
    subject: e.subject,
    start: e.start?.dateTime?.slice(0, 16),
    end: e.end?.dateTime?.slice(0, 16),
    location: e.location?.displayName,
    organizer: e.organizer?.emailAddress?.address,
    isAllDay: e.isAllDay,
    showAs: e.showAs,
    isOnlineMeeting: e.isOnlineMeeting,
    joinUrl: e.onlineMeeting?.joinUrl,
  })));
}

async function cmdSearch(args) {
  const query = args._[0];
  if (!query) { console.error('Usage: teams search "<query>"'); process.exit(1); }
  
  // Use /search/query endpoint
  const result = await graphPost('/search/query', {
    requests: [{
      entityTypes: ['chatMessage'],
      query: { queryString: query },
      from: 0,
      size: 25,
    }],
  });
  
  const hits = result.value?.[0]?.hitsContainers?.[0]?.hits || [];
  out(hits.map(h => ({
    summary: h.summary,
    from: h.resource?.from?.emailAddress?.name,
    date: h.resource?.createdDateTime,
    body: h.resource?.body?.content?.replace(/<[^>]*>/g, '').slice(0, 200),
  })));
}

// ─── Router ───
const args = parseArgs(process.argv.slice(2));
const [cmd, sub, ...rest] = args._;
const subArgs = { _: sub ? [sub, ...rest] : rest, flags: args.flags };

try {
  switch (cmd) {
    case 'token':
      if (sub === 'extract-browser') await cmdTokenExtractBrowser();
      else if (sub === 'store') await cmdTokenStore({ _: rest, flags: args.flags });
      else if (sub === 'test') await cmdTokenTest();
      else console.error('Usage: teams token [extract-browser|store|test]');
      break;
    case 'me': await cmdMe(); break;
    case 'teams': await cmdTeams(); break;
    case 'chats': await cmdChats(subArgs); break;
    case 'chat': await cmdChatMessages(subArgs); break;
    case 'chat-send': await cmdChatSend(subArgs); break;
    case 'channels': await cmdChannels(subArgs); break;
    case 'channel': await cmdChannelMessages(subArgs); break;
    case 'channel-send': await cmdChannelSend(subArgs); break;
    case 'presence': await cmdPresence(subArgs); break;
    case 'users': await cmdUsers(subArgs); break;
    case 'activity': await cmdActivity(subArgs); break;
    case 'calendar': await cmdCalendar(subArgs); break;
    case 'search': await cmdSearch(subArgs); break;
    default:
      console.log(`teams — Microsoft Teams via Graph API

Token:
  token extract-browser              Print JS to extract token from Teams tab
  token store --refresh-token <tok>  Store extracted token
  token test                         Verify token works

Messaging:
  chats [--top 20]                   List recent chats
  chat <chatId> [--top 30]           Read chat messages
  chat-send <chatId> --message <txt> Send message to chat
  search "<query>"                   Search messages

Teams & Channels:
  teams                              List joined teams
  channels <teamId>                  List channels
  channel <teamId> <channelId>       Read channel messages
  channel-send <teamId> <channelId> --message <txt>

People & Status:
  me                                 Your profile
  users [--search X] [--top 25]      Search org directory
  presence [--user-id <id>]          Presence status

Calendar:
  calendar [--days 7]                Upcoming meetings

🔗 Shares token with Outlook skill (~/.openclaw/credentials/outlook-msal.json)`);
      break;
  }
} catch (e) {
  console.error(`❌ ${e.message}`);
  process.exit(1);
}

File: .skills/openclaw-skills/skills/takeovernat/teams-delegate/scripts/teams.py
#!/usr/bin/env python3
"""
teams-delegate core CLI

Usage:
  python3 teams.py whoami          # Get your user ID (run this first)
  python3 teams.py inbox           # List recent chats
  python3 teams.py read <chatId>   # Show messages in a chat
  python3 teams.py reply <chatId> "message"
  python3 teams.py summary         # AI-ready summary of all chats
"""

import json, re, sys, urllib.request, urllib.parse, urllib.error
from datetime import datetime

sys.path.insert(0, __file__.rsplit("/", 1)[0])
from auth import get_token, GRAPH_BASE

CONFIG_DIR = __file__.rsplit("/", 2)[0]

def graph_get(path, token):
    url = f"{GRAPH_BASE}{path}"
    req = urllib.request.Request(url, headers={
        "Authorization": f"Bearer {token}",
        "Accept": "application/json"
    })
    try:
        with urllib.request.urlopen(req) as r:
            return json.loads(r.read())
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        print(f"Graph API error {e.code} on {path}:", file=sys.stderr)
        try:
            err = json.loads(body)
            print(f"  {err.get('error', {}).get('code')}: {err.get('error', {}).get('message')}", file=sys.stderr)
        except:
            print(f"  {body[:300]}", file=sys.stderr)
        sys.exit(1)

def graph_post(path, token, payload):
    url = f"{GRAPH_BASE}{path}"
    req = urllib.request.Request(url,
        data=json.dumps(payload).encode(),
        headers={"Authorization": f"Bearer {token}",
                 "Content-Type": "application/json",
                 "Accept": "application/json"},
        method="POST")
    with urllib.request.urlopen(req) as r:
        return json.loads(r.read())

def strip_html(text):
    return re.sub(r"<[^<]+?>", "", text or "").strip()

def fmt_time(iso):
    try:
        dt = datetime.fromisoformat(iso.replace("Z", "+00:00"))
        return dt.astimezone().strftime("%m/%d %I:%M %p")
    except:
        return iso

def get_user_id(token):
    # For app-only tokens, list users and pick first
    # For delegated tokens, use /me
    try:
        me = graph_get("/me", token)
        return me["id"], me.get("displayName", "You")
    except SystemExit:
        # App-only token — need to list users
        users = graph_get("/users?$top=5", token)
        if not users.get("value"):
            print("No users found. Make sure User.Read.All permission is granted.", file=sys.stderr)
            sys.exit(1)
        print("\nAvailable users (app-only mode):")
        for i, u in enumerate(users["value"]):
            print(f"  {i+1}. {u.get('displayName')} — {u.get('mail') or u.get('userPrincipalName')} (id: {u['id']})")
        choice = input("\nWhich user? Enter number: ").strip()
        user = users["value"][int(choice)-1]
        return user["id"], user.get("displayName", "User")

def get_chats(user_id, token):
    return graph_get(f"/users/{user_id}/chats?$expand=lastMessagePreview&$top=20", token).get("value", [])

def get_messages(user_id, chat_id, token, top=10):
    return graph_get(f"/users/{user_id}/chats/{chat_id}/messages?$top={top}", token).get("value", [])

def send_reply(user_id, chat_id, message, token):
    return graph_post(f"/users/{user_id}/chats/{chat_id}/messages", token,
                      {"body": {"contentType": "text", "content": message}})

def cmd_whoami(token):
    uid, name = get_user_id(token)
    print(f"User: {name}\nID:   {uid}")

def cmd_inbox(token):
    uid, name = get_user_id(token)
    chats = get_chats(uid, token)
    if not chats:
        print("No recent chats."); return
    print(f"\nInbox for {name}\n{'─'*80}")
    for c in chats:
        chat_id   = c["id"]
        chat_type = c.get("chatType", "?")
        preview   = c.get("lastMessagePreview", {})
        sender    = preview.get("from", {}).get("user", {}).get("displayName", "?")
        content   = strip_html(preview.get("body", {}).get("content", ""))[:50]
        t         = fmt_time(preview.get("createdDateTime", ""))
        print(f"[{t}] {chat_type:<10} {sender}: {content}")
        print(f"  ID: {chat_id}\n")

def cmd_read(chat_id, token):
    uid, _ = get_user_id(token)
    msgs = get_messages(uid, chat_id, token)
    msgs.reverse()
    print(f"\nChat: {chat_id}\n{'='*60}")
    for m in msgs:
        sender  = m.get("from", {}).get("user", {}).get("displayName", "?")
        content = strip_html(m.get("body", {}).get("content", ""))
        t       = fmt_time(m.get("createdDateTime", ""))
        flag    = " [URGENT]" if m.get("importance") == "urgent" else ""
        print(f"\n[{t}]{flag} {sender}:\n  {content}")

def cmd_summary(token):
    uid, name = get_user_id(token)
    chats = get_chats(uid, token)
    print(f"\n=== TEAMS INBOX SUMMARY — {name} ===\n")
    for c in chats[:10]:
        preview = c.get("lastMessagePreview", {})
        sender  = preview.get("from", {}).get("user", {}).get("displayName", "?")
        content = strip_html(preview.get("body", {}).get("content", ""))
        t       = fmt_time(preview.get("createdDateTime", ""))
        imp     = preview.get("importance", "normal")
        print(f"CHAT_ID: {c['id']}")
        print(f"  Type: {c.get('chatType')} | From: {sender} | Time: {t} | Importance: {imp}")
        print(f"  Preview: {content[:120]}\n")

def cmd_reply(chat_id, message, token):
    uid, _ = get_user_id(token)
    result = send_reply(uid, chat_id, message, token)
    print(f"Reply sent (id: {result.get('id')})")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(__doc__); sys.exit(1)
    token = get_token()
    cmd   = sys.argv[1]
    if   cmd == "whoami":                          cmd_whoami(token)
    elif cmd == "inbox":                           cmd_inbox(token)
    elif cmd == "read"   and len(sys.argv) >= 3:  cmd_read(sys.argv[2], token)
    elif cmd == "reply"  and len(sys.argv) >= 4:  cmd_reply(sys.argv[2], sys.argv[3], token)
    elif cmd == "summary":                         cmd_summary(token)
    else: print(__doc__); sys.exit(1)
